File: server/src/backup.ts
import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { readFile, unlink } from 'fs/promises';
import { join } from 'path';
import { tmpdir } from 'os';
import { config } from './config';
import { db } from './db';
import { log } from './logger';

export async function runBackup(): Promise<void> {
  if (!config.backup_bucket) {
    log.debug('backup: no backup_bucket configured, skipping');
    return;
  }

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const filename = `cacher-${timestamp}.db`;
  const tmpPath = join(tmpdir(), filename);

  try {
    log.debug(`backup: writing snapshot to ${tmpPath}`);
    db.exec(`VACUUM INTO '${tmpPath.replace(/'/g, "''")}'`);

    const data = await readFile(tmpPath);

    const client = new S3Client({
      region: 'auto',
      endpoint: `https://${config.r2.account_id}.r2.cloudflarestorage.com`,
      credentials: {
        accessKeyId: config.r2.access_key_id,
        secretAccessKey: config.r2.secret_access_key,
      },
    });

    await client.send(new PutObjectCommand({
      Bucket: config.backup_bucket,
      Key: filename,
      Body: data,
      ContentType: 'application/octet-stream',
    }));

    log.info(`backup: uploaded ${filename} (${data.length} bytes) to ${config.backup_bucket}`);
  } catch (err) {
    log.error(`backup: failed — ${err}`);
  } finally {
    await unlink(tmpPath).catch(() => {});
  }
}

File: goserver/internal/backup.go
package internal

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func RunBackup() error {
	if Cfg.BackupBucket == "" {
		LogDebug("backup: no backup_bucket configured, skipping")
		return nil
	}

	timestamp := time.Now().UTC().Format("2006-01-02T15-04-05")
	filename := "cacher-" + timestamp + ".db"
	tmpPath := filepath.Join(os.TempDir(), filename)
	defer os.Remove(tmpPath)

	LogDebug("backup: writing snapshot to %s", tmpPath)
	escapedPath := strings.ReplaceAll(tmpPath, "'", "''")
	if _, err := DB.Exec(fmt.Sprintf("VACUUM INTO '%s'", escapedPath)); err != nil {
		return fmt.Errorf("vacuum into: %w", err)
	}

	data, err := os.ReadFile(tmpPath)
	if err != nil {
		return fmt.Errorf("read snapshot: %w", err)
	}

	client := NewBackupS3Client()
	_, err = client.PutObject(context.Background(), &s3.PutObjectInput{
		Bucket:      aws.String(Cfg.BackupBucket),
		Key:         aws.String(filename),
		Body:        bytes.NewReader(data),
		ContentType: aws.String("application/octet-stream"),
	})
	if err != nil {
		return fmt.Errorf("upload: %w", err)
	}

	LogInfo("backup: uploaded %s (%d bytes) to %s", filename, len(data), Cfg.BackupBucket)
	return nil
}
